job/strategies/templates/strategy.py

```diff
@@ -1,57 +1,103 @@
 from __future__ import annotations
 
+import logging
+import time
 from abc import ABCMeta, abstractmethod
-from typing import TYPE_CHECKING, Any, Dict, List
+from typing import TYPE_CHECKING, List
 
 import numpy as np
 import pandas as pd
 
+from db.helpers import create_model, refresh_db, set_current_model
 from db.mappings.model import ModelType
 from db.mappings.recommendation import Rec
+from job.helpers.itertools import batch
+from job.helpers.knn import KNN, map_neighbors_to_recommendations
+from lib.config import config
 
 if TYPE_CHECKING:
     from sites.templates.site import Site
 
+MAX_RECS = config.get("MAX_RECS")
+
 
 class Strategy(metaclass=ABCMeta):
     """
     Superclass that defines the methods for each one of the recommendation strategies (collaborative filtering,
     popularity, semantic similarity).
     """
 
+    decays: np.ndarray
+    train_embeddings: np.ndarray
+    recommendations: List[Rec]
+    model_type: ModelType = None
+    train_data: pd.DataFrame = None
+
     @abstractmethod
-    def fetch_data(self, site: Site, interactions_data: pd.DataFrame) -> List[Dict[str, Any]]:
+    def fetch_data(self, site: Site, interactions_data: pd.DataFrame = None, experiment_time=None) -> None:
```

Comment:
The problem with adding `experiment_time` to `__init__` is that the constructor is executed in places like this: [`philadelphia_inquirer.py`](https://github.com/LocalAtBrown/article-rec-training-job/blob/ea40e14d47b4e13f6c79d4aea90aba57ff722d51/sites/philadelphia_inquirer.py#L281)

And `experiment_time` is ultimately `EXPERIMENT_DT` from [`job.py`](https://github.com/LocalAtBrown/article-rec-training-job/blob/ea40e14d47b4e13f6c79d4aea90aba57ff722d51/job/job.py#L24).

Therefore the constructor is executed before `EXPERIMENT_DT` is defined.

Could we have different `experiment_time` for different strategies in the same site? Let's say we run **collaborative filtering** and then **popularity** for the Texas Tribune. Is it a problem that `experiment_time` for **popularity** is a few minutes later than for **collaborative filtering**?

1. If that's not a problem, we could add something like this to the `__init__` of `strategy.py`:

`self.experiment_time = datetime.now()`

If all strategies need to have the same `experiment_time`, we have two alternatives:

2. Refactor `Sites` and `Site` to receive a parameter called `experiment_time` that will then be transmitted to its `Strategy` attributes.
3. We continue to have `experiment_time` as an argument in `fetch_data`.

I prefer (1) and would be fine with (3). But I really dislike (2) because I think the code we inherited is already very convoluted.
What do you think?

Reply:
Ah! Good call. Then I think it's okay to keep `self.experiment_time = experiment_time` outside `__init__`. Maybe we can put that inside its own abstract method with a name like `set_experiment_time`? My preference would be to separate it from `fetch_events` because it's not really part of the additional-data-fetching logic.